titan-compiler/coder.lua

```diff
@@ -761,6 +761,85 @@ generate_var = function(var)
     end
 end
 
+local function generate_exp_builtin_table_insert(exp)
+    local args = exp.args
+    assert(#args == 2)
+    local cstats_t, cvalue_t = generate_exp(args[1])
+    local cstats_v, cvalue_v = generate_exp(args[2])
+    local ui = tmp_name()
+    local ui_decl = c_declaration("lua_Unsigned", ui)
+    local size = tmp_name()
+    local size_decl = c_declaration("lua_Unsigned", size)
+    local slot = tmp_name()
+    local slot_decl = c_declaration("TValue*", slot)
+    local cstats = util.render([[
+        ${CSTATS_T}
+        ${CSTATS_V}
+        ${UI_DECL} = luaH_getn(${CVALUE_T});
+        ${SIZE_DECL} = ${CVALUE_T}->sizearray;
+        if (${SIZE} <= ${UI}) {
+            if (${SIZE} < 8) {
+                /* avoid infinite loop if sizearray == 0 */
+                ${SIZE} = 8;
+            }
+            while (${SIZE} <= ${UI}) {
```

Comment:
There is no other way to compute this without a loop?

Reply:
I think this loop should only ever need to go through one iteration but am not 100% sure about that so I played safe and went with the while loop

Reply:
Couldn't this while be replaced with an else? If we increment the array by one, doubling the size should always work.